### ISSUE TITLE
What is the time that a task should be saved?

### ISSUE THREAD

**Issue by Jacky142857** [nus-cs2103-AY2122S2/forum#38]
For the save function, can I save the task when I terminate the application (when I type "bye"), or do I need to save the task whenever a command is being executed ?

**Comment by tonishka** [comment]
Hello, 

The way I did it was - whenever the user starts the program the data is loaded from the file to an ArrayList. Throughout the execution of the program the data manipulation takes place in the ArrayList and when the user types the exit command the data from the ArrayList is written back to the file. Writing back to file after every command may be too cumbersome, it is easier to write back to file in one go after the program has finished execution.

TL;DR: I save the data on termination.

**Comment by Spyobird** [comment]
Writing on each execution may be 'slower' and 'more inefficient', however, writing on termination may not result in the changes being saved in the event of unexpected termination of the program (e.g. crashes, ctrl+C). It is possible to handle those cases through error handling though.

**Comment by jbkim1999** [comment]
What I did was for an every operation that involved modifying the TaskList (e.g. adding a task or marking/unmarking), my Storage class would evoke save method and save the result back to the txt file. But now that I think of it, I believe I could only save the file for bye command, but like @Spyobird mentioned, sudden termination must be handled in such case.

**Comment by ddx-510** [comment]
Although both ways would be fine, but from the IP task description on **save**

> Save the tasks in the hard disk automatically whenever the task list changes. 

It means we need a `save` operation whenever there is a modification on `TaskList`, hence we should `save` after every command that modifies the `TaskList` (`mark`, `unmark`, `delete`, etc.)

**Comment by albertsutz** [comment]
I would say that you can add both if you are not sure. In the end, the functionality / the core idea of the method is the same. You can just automatically save after "bye" and save deliberately when the user enter "save".

**Comment by Jacky142857** [comment]
I found that implementing save between every command makes my application slower. Though the guide instructs us to save for every command line, I think maybe it is better to save it asynchronously.

**Comment by damithc** [comment]
> I found that implementing save between every command makes my application slower. Though the guide instructs us to save for every command line, I think maybe it is better to save it asynchronously.

@Jacky142857 how much slower?

**Comment by Jacky142857** [comment]
> 

Sorry it is about the same speed, maybe I added my subjective view when looking at the speed

**Comment by damithc** [comment]
> Sorry it is about the same speed, maybe I added my subjective view when looking at the speed

@Jacky142857 You are not the only one :-) Most programmers would instinctively look for a more 'efficient' solution even if the gains are insignificant and costs (e.g., extra implementation effort, extra complexity in the code) outweigh them.

More on this https://wiki.c2.com/?PrematureOptimization
